rasa/utils/tensorflow/layers.py

```diff
@@ -1085,3 +1140,314 @@ def call(
+
+        accuracy = self._accuracy(sim_pos, sim_candidate_il, pos_neg_labels)
+        loss = self._loss_sigmoid(sim_pos, sim_candidate_il, pos_neg_labels, mask)
+
+        return loss, accuracy
+
+    def _sample_candidates(
+        self,
+        batch_inputs_embed: tf.Tensor,  # (batch_size, 1, num_features)
+        batch_labels_embed: tf.Tensor,  # (batch_size, max_num_labels_per_input, num_features)
+        batch_labels_ids: tf.Tensor,  # (batch_size, max_num_labels_per_input, 1)
+        all_labels_embed: tf.Tensor,  # (num_labels, num_features)
+        all_labels_ids: tf.Tensor,  # (num_labels, 1)
+    ) -> Tuple[
+        tf.Tensor,  # (batch_size, 1, 1, num_features)
+        tf.Tensor,  # (batch_size, 1, num_features)
+        tf.Tensor,  # (batch_size, 1, num_candidates, num_features)
+        tf.Tensor,  # (batch_size, num_candidates)
+    ]:
+        """Sample candidate examples.
+        
+        Args:
+            batch_inputs_embed: Embeddings of the batch inputs (e.g. featurized trackers)
+            batch_labels_embed: Embeddings of the batch labels (e.g. featurized intents for IntentTED)
+            batch_labels_ids: Batch label indices (e.g. indices of the intents)
+            all_labels_embed: Embeddings for all labels in the domain
+            all_labels_ids: Indices for all labels in the domain
+
+        Returns:
+            pos_inputs_embed: Embeddings of the batch inputs
+            pos_labels_embed: First example of the embedding of a positive label
+            candidate_labels_embed: More examples of embeddings of labels, some positive some negative
+            pos_neg_labels: Indicator for which candidates are positives and which are negatives
+        """
+        pos_inputs_embed = tf.expand_dims(
+            batch_inputs_embed, axis=-2, name="expand_pos_input"
+        )
+
+        # We want to guarantee that we return at least one positive example. All labels
+        # in `batch_labels_embed` are positive examples, but their number can be different
+        # in each example. So we take the first positive one here as our guarantee, and
+        # we may or may not capture more positive examples with the candidate sampling below.
+        pos_labels_embed = tf.expand_dims(
+            tf.expand_dims(batch_labels_embed[:, 0, ...], axis=-2),
+            axis=1,
+            name="expand_pos_labels",
+        )
+
+        # Pick random examples from the batch
+        candidate_ids = self._get_candidate_indices(
+            batch_size=tf.shape(batch_inputs_embed)[0],
+            total_candidates=tf.shape(all_labels_embed)[0],
+        )
+
+        # Get the label embeddings and ids corresponding to candidate indices
+        candidate_labels_embed = self._get_candidate_values(
+            all_labels_embed, candidate_ids
+        )
+        candidate_labels_ids = self._get_candidate_values(all_labels_ids, candidate_ids)
+
+        # Determine how many distinct labels exist (highest label index)
```

Comment:
The most general way would be to convert `all_labels_ids` into a python set and measure its lengh. But this would be slow and move outside the TF graph. I think the way it is done now is good. If we ever use strings or other objects as label IDs, much of the other code wouldn't work either.